chore: Fix solid queue tests

adamlogic opened the pull request:
Not sure yet if it's just a test issue with recent Solid Queue versions, or if it's an actual compatibility issue.

adamlogic committed:
Temporarily restrict solid_queue version to get specs passing

adamlogic committed:
Move past the ActiveModel error

New error in Solid Queue 0.8.1 seems related to how we setup our test DB.

adamlogic commented:
@carlosantoniodasilva I'm incrementally moving through Solid Queue versions to see where things are failing. Our Solid Queue tests started failing in v0.7.1 with `uninitialized constant SolidQueue::Processes::Callbacks::ActiveModel`. I've fixed that by requiring it explicitly, although I'm unclear why we need to do that. 🤷‍♂️ 

Now we're [encountering DB issues](https://github.com/judoscale/judoscale-ruby/actions/runs/11307557038/job/31449409249?pr=222) in v0.8.1 where [there were substantial changes](https://github.com/rails/solid_queue/releases/tag/v0.8.0) to how Solid Queue prepares the schema. I'm a bit fuzzy on what's happening in our `test_helper` to bootstrap the Solid Queue schema, so I'm hoping you could jump in and take a look.

carlosantoniodasilva commented on judoscale-solid_queue/lib/judoscale/solid_queue.rb:
```diff
@@ -1,5 +1,8 @@
 
+# Tests are failing with solid_queue 0.7.1+ unless we require active_model.
+# Not sure why since solid_queue depends on activerecord which depends on activemodel.
+require "active_model"
```
They're not explicitly requiring `active_model` (and/or the specific [callbacks module that they rely on](https://github.com/rails/solid_queue/blob/51c75bec01c8c53f2fe573822f0158e8b62030aa/lib/solid_queue/processes/callbacks.rb#L8)), so I'm guessing by the time that `solid_queue` code gets loaded, it hasn't been required by AR yet.

I was hoping to fix it there, but if we want to support prior versions we might have to keep a require on our side for the time being as well.

adamlogic replied:
Ah, got it. 👍

carlosantoniodasilva reviewed (commented):
Thanks for getting it started, I'll take a look at those schema changes.

Do we want to support multiple versions, or maybe we should lock on v1+ once we get tests running there?

adamlogic commented:
> Do we want to support multiple versions, or maybe we should lock on v1+ once we get tests running there?

Replying in Linear.

carlosantoniodasilva committed:
Load entire SolidQueue schema file manually into our test DB

SolidQueue v0.8+ merged all the migrations into a single schema file
that is automatically loaded onto the DB on a Rails app. They suggest a
separate DB, which AR will handle automatically by copying over this
single file schema to the app, and to use on the same DB one would have
to manually move the schema to a migration and execute it.

For our case, all we care about is to have SolidQueue schema / tables
loaded onto our test DB. We do that by manually loading the lib's schema
file via Active Record. We still execute migrations afterwards, even
though it's a no-op for current SolidQueue, if they need to make new
schema changes in the future, they'll come in the form of migrations.

carlosantoniodasilva committed:
Merge branch 'main' into fix-solid-queue-tests

carlosantoniodasilva committed:
Revert dependency check to continue support older versions

carlosantoniodasilva committed:
Add Gemfile to test SolidQueue minimum version 0.3.x

This way we can test both minimum and whatever "current/latest" is.
(currently v1.0)

carlosantoniodasilva committed:
Add specific Gemfile for judoscale-rails against v7.1

Update GH matrix to skip unsupported Ruby versions on Rails 7.2 as well,
(main Gemfile for now) since it requires Ruby 3.1+.

carlosantoniodasilva committed:
Move explicit require of ActiveModel to the test helper

We already have a few other requires that were needed there to get tests
running, we can keep this new one there too, as it should not be a
problem in runtime within a normal Rails environment. (and if it happens
to be, we'll learn about it soon enough.)

carlosantoniodasilva commented on .github/workflows/judoscale-rails-test.yml:
```diff
@@ -14,6 +14,7 @@ jobs:
       matrix:
         gemfile:
           - Gemfile
+          - Gemfile-rails-7-1
```
Adding a new Gemfile to test Rails 7.1, keeping the main Gemfile testing 7.2 (8.0 still in alpha)

carlosantoniodasilva commented on .github/workflows/judoscale-solid_queue-test.yml:
```diff
@@ -14,6 +14,7 @@ jobs:
       matrix:
         gemfile:
           - Gemfile
+          - Gemfile-solid_queue-0-3
```
New Gemfile to test against solid_queue v0.3 so we get coverage on minimum version supported

carlosantoniodasilva commented on judoscale-solid_queue/test/test_helper.rb:
```diff
@@ -9,6 +9,9 @@
 require "active_support/core_ext/numeric/time"
+# It seems SolidQueue v0.7.1+ blows up without requiring activemodel.
+# It should be a dependency for them through activerecord, but they may not be requiring it properly.
+require "active_model"
```
Moved the require of active_model to the test helper, as with a few others above. I don't think this should be a problem within a running app, hopefully.

carlosantoniodasilva commented on judoscale-solid_queue/test/test_helper.rb:
```diff
@@ -41,10 +44,15 @@ class TestRailsApp < Rails::Application
 }
-ActiveRecord::Base.establish_connection(DATABASE_URL)
+ActiveRecord::Base.configurations = {test: {url: DATABASE_URL}}
+ActiveRecord::Base.establish_connection(:test)
```
Need to establish the connection via the current Rails.env to `load_schema_current` below.

carlosantoniodasilva commented on judoscale-solid_queue/test/test_helper.rb:
```diff
@@ -41,10 +44,15 @@ class TestRailsApp < Rails::Application
+# SolidQueue v0.8+ merged migrations into a single schema file, which we load directly into our test DB.
+# Migrations are still executed afterwards, as they may add them in the future if schema changes are needed.
+SCHEMA_FILE = SolidQueue::Engine.config.paths["lib"].paths.first.join("generators", "solid_queue", "install", "templates", "db", "queue_schema.rb")
+ActiveRecord::Tasks::DatabaseTasks.load_schema_current(ActiveRecord.schema_format, SCHEMA_FILE) if SCHEMA_FILE.exist?
```
Load the full SolidQueue schema from the gem if available.

carlosantoniodasilva reviewed (commented):
@adamlogic this should be green now, let me know what you think.

adamlogic commented:
@carlosantoniodasilva Awesome, thank you!
